### ISSUE TITLE
group-by with aggregate functions not supported in database backends?

### ISSUE THREAD

**Issue by javierluraschi** [tidyverse/dplyr#2887]
`source %>% group_by(column) %>% mutate(column_sd = sd(column))` does not seem to be implemented for database sources.

### Data Frames

With data frames we can run:

```r
iris %>%
  group_by(Species) %>%
  mutate(A_sd = sd(Sepal.Length))
```

```
# A tibble: 150 x 6
# Groups:   Species [3]
   Sepal.Length Sepal.Width Petal.Length Petal.Width Species      A_sd
          <dbl>       <dbl>        <dbl>       <dbl>  <fctr>     <dbl>
 1          5.1         3.5          1.4         0.2  setosa 0.3524897
 2          4.9         3.0          1.4         0.2  setosa 0.3524897
 3          4.7         3.2          1.3         0.2  setosa 0.3524897
...
```

However, I have not seen this implemented in any database backend.

### RPostgreSQL

But no `dbplyr` backend seems to support this, from `RPostgreSQL`:

```
p <- src_postgres(host = "", port = "", dbname = "", user = "", password = "")
copy_to(p, iris)

tbl(p, "iris") %>%
  group_by(Species) %>%
  mutate(A_sd = sd(Sepal.Length))
```

```
# Source:   lazy query [?? x 6]
# Database: postgres 9.4.7
#   [rstudioadmin@sol-eng-postgre.cihykudhzbgw.us-west-2.rds.amazonaws.com:5432/airontime]
# Groups:   Species
   Sepal.Length Sepal.Width Petal.Length Petal.Width Species  A_sd
          <dbl>       <dbl>        <dbl>       <dbl>   <chr> <chr>
 1          5.1         3.5          1.4         0.2  setosa  <NA>
 2          4.9         3.0          1.4         0.2  setosa  <NA>
 3          4.7         3.2          1.3         0.2  setosa  <NA>
...
```

```r
tbl(p, "iris") %>% group_by(Sepal.Length) %>% mutate(A_sd = sd(Sepal.Width)) %>% sql_render()
```

**Currently:**

```
<SQL> SELECT "Sepal.Length", "Sepal.Width", "Petal.Length", "Petal.Width", "Species", NULL AS "A_sd"
FROM "iris"
```

**Expected:**

```
tbl(p, sql("SELECT \"Sepal.Length\", \"Sepal.Width\", \"Petal.Length\", \"Petal.Width\", \"Species\", a_sd FROM (SELECT l.*, l.\"Species\" as ls, r.\"Species\" as rs, a_sd A_sd FROM iris as l, (SELECT \"Species\", stddev_samp(\"Sepal.Length\") as A_sd FROM iris GROUP BY \"Species\") r) m WHERE ls = rs"))
```

```
# Source:   SQL [?? x 6]
# Database: postgres 9.4.7
#   [rstudioadmin@sol-eng-postgre.cihykudhzbgw.us-west-2.rds.amazonaws.com:5432/airontime]
   Sepal.Length Sepal.Width Petal.Length Petal.Width Species      a_sd
          <dbl>       <dbl>        <dbl>       <dbl>   <chr>     <dbl>
 1          5.1         3.5          1.4         0.2  setosa 0.3524897
 2          4.9         3.0          1.4         0.2  setosa 0.3524897
 3          4.7         3.2          1.3         0.2  setosa 0.3524897
```

### sparklyr

```r
spark_install(version = "2.1.0")
sc <- spark_connect(master = "local", version = "2.1.0")
iris_tbl <- copy_to(sc, iris)

iris_tbl %>%
  group_by(Species) %>%
  mutate(A_sd = sd(Sepal.Length))
```

```
# Source:   lazy query [?? x 6]
# Database: spark_connection
# Groups:   Species
   Sepal_Length Sepal_Width Petal_Length Petal_Width Species  A_sd
          <dbl>       <dbl>        <dbl>       <dbl>   <chr> <chr>
 1          5.1         3.5          1.4         0.2  setosa  <NA>
 2          4.9         3.0          1.4         0.2  setosa  <NA>
 3          4.7         3.2          1.3         0.2  setosa  <NA>
...
```

```
iris_tbl %>%
  group_by(Species) %>%
  mutate(A_sd = sd(Sepal.Length)) %>% sql_render()
```

**Currently:**

```
<SQL> SELECT `Sepal_Length`, `Sepal_Width`, `Petal_Length`, `Petal_Width`, `Species`, NULL AS `A_sd`
FROM `iris`
```

**Expected:**

```
<SQL> SELECT l.*, A_sd FROM iris as l LEFT JOIN (SELECT Species, stddev_samp(Sepal_Length) as A_sd FROM iris GROUP BY Species) r ON l.Species = r.Species
```

```
# Source:   SQL [?? x 6]
# Database: spark_connection
   Sepal_Length Sepal_Width Petal_Length Petal_Width Species      A_sd
          <dbl>       <dbl>        <dbl>       <dbl>   <chr>     <dbl>
 1          5.1         3.5          1.4         0.2  setosa 0.3524897
 2          4.9         3.0          1.4         0.2  setosa 0.3524897
 3          4.7         3.2          1.3         0.2  setosa 0.3524897
```

## Workarounds ##

While we've recommended using the following workaround, users still expect the compressed version to work as expected.

```r
iris_tbl %>%
  group_by(Species) %>%
  summarize(A_sd = sd(Sepal_Length)) %>% left_join(iris_tbl)
```

```
# Source:   lazy query [?? x 6]
# Database: spark_connection
      Species      A_sd Sepal_Length Sepal_Width Petal_Length Petal_Width
        <chr>     <dbl>        <dbl>       <dbl>        <dbl>       <dbl>
 1 versicolor 0.5161711          5.7         2.8          4.1         1.3
 2 versicolor 0.5161711          5.1         2.5          3.0         1.1
 3 versicolor 0.5161711          6.2         2.9          4.3         1.3
```

**Comment by hadley** [comment]
Would you mind making a reprex with memdb_frame? I'll take a look tomorrow.

**Comment by javierluraschi** [comment]
Yes, for sure:

```r
iris_mem <- copy_to(src_memdb(), iris)
iris_mem %>%
  group_by(Species) %>%
  mutate(A_sd = sd(Sepal.Length))
```

**Actual:**

```
# Source:   lazy query [?? x 6]
# Database: sqlite 3.19.3 [:memory:]
# Groups:   Species
   Sepal.Length Sepal.Width Petal.Length Petal.Width Species  A_sd
          <dbl>       <dbl>        <dbl>       <dbl>   <chr> <lgl>
 1          5.1         3.5          1.4         0.2  setosa    NA
 2          4.9         3.0          1.4         0.2  setosa    NA
 3          4.7         3.2          1.3         0.2  setosa    NA
```

**Expected:**

```
dbGetQuery(iris_mem$src$con, "SELECT l.*, A_sd FROM iris as l LEFT JOIN (SELECT Species, stdev(\"Sepal.Length\") as A_sd FROM iris GROUP BY Species) r ON l.Species = r.Species")
```

```
    Sepal.Length Sepal.Width Petal.Length Petal.Width    Species      A_sd
1            5.1         3.5          1.4         0.2     setosa 0.3524897
2            4.9         3.0          1.4         0.2     setosa 0.3524897
3            4.7         3.2          1.3         0.2     setosa 0.3524897
4            4.6         3.1          1.5         0.2     setosa 0.3524897
```

**Comment by edgararuiz** [comment]
@hadley / @javierluraschi  - I just added window support of `sd()` for all  sources in this PR: https://github.com/tidyverse/dbplyr/pull/20

I also did my best to add scalar and agg support of `sd()` for those that didn't have it.  I couldn't find a reference of SD for SQL92, so I didn't touch the base translations.

Here's the new result from a MSSQL connection: 


```
> tbl(con, "iris") %>%
+   group_by(Species) %>%
+   mutate(A_sd = sd(Sepal.Length))
```

```
# Source:   lazy query [?? x 6]
# Database: Microsoft SQL Server 12.00.4422[rstudioadmin@EC2AMAZ-O5QKKM8/airontime]
# Groups:   Species
   Sepal.Length Sepal.Width Petal.Length Petal.Width Species      A_sd
          <dbl>       <dbl>        <dbl>       <dbl>   <chr>     <dbl>
 1          5.1         3.5          1.4         0.2  setosa 0.3524897
 2          4.9         3.0          1.4         0.2  setosa 0.3524897
 3          4.7         3.2          1.3         0.2  setosa 0.3524897
```

**Comment by hadley** [comment]
@edgararuiz thanks! 

@javierluraschi dplyr doesn't know how to turn recycled window functions into a left join if the database doesn't support window functions.

I think there's still a bug here because the original code should throw an error. I'll take a look on Thursday

**Comment by javierluraschi** [comment]
@hadley Spark/`sparklyr` does support window functions.... https://databricks.com/blog/2015/07/15/introducing-window-functions-in-spark-sql.html - For instance, I know that `row_number`, `ntile`, `min_rank`, `dense_rank`, `percent_rank` and `cume_dist` work appropiately with `dplyr` and `sparklyr`.

Is there something that you would expect a database to implement? Sounds like there is, but I'm not following you here. How would the query look like and which `dplyr` endpoints need to be implemented to support that? Thanks!

**Comment by hadley** [comment]
@javierluraschi SQLite doesn't have window functions, so it should definitely be throwing an error.  And for any database, there's no way it should be generating "NULL" as the SQL translation.

**Comment by hadley** [comment]
Minimal reprex:

```R
library(dplyr)
translate_sql(sd(x), con = src_memdb()$con)
#> Warning in is.data.frame(x) : NAs introduced by coercion
#> <SQL> NULL
```

**Comment by hadley** [comment]
Even more minimal reprex:

```R
translate_sql(var(x))
#> Warning in var(x) : NAs introduced by coercion
#> <SQL> NULL
```

**Comment by hadley** [comment]
Ok, this problem occurs if you've defined a aggregation function but not the equivalent window function. I'll add a check for this issue in `sql_variant()`.

**Comment by hadley** [comment]
Fixed in tidyverse/dbplyr@76b6801e8b6ad90735ee636c48e576e3d74d2813
